Feature/tighten hash usage 2 (fixing Issues #316, #423, and probably #492)

JoshuaGreen opened the pull request:
This branch is an implementation of the same changes as those in [`feature/tighten_hash_usage`](https://github.com/thomas-maeder/popeye/tree/feature/tighten_hash_usage) but based off of [`develop` commit #5ffb472](https://github.com/thomas-maeder/popeye/commit/5ffb4723d8c608553ecaf6d15b114883f7b7d544).&nbsp; See PR #441 for the full explanation of what's going on here.

I believe this resolves Issues #316 and #423.&nbsp; Additionally, I assume it resolves Issue #492, though as per the discussion there one could take the relevant changes even further if desired.

MuTsunTsai commented:
I saw you left DOMEASURE on in this PR. Shouldn't it be off by default?

JoshuaGreen commented:
@MuTsunTsai, that's set in [`develop` commit #5ffb472](https://github.com/thomas-maeder/popeye/commit/5ffb4723d8c608553ecaf6d15b114883f7b7d544), so I left it alone.&nbsp; It's useful for testing, and I trust @thomas-maeder to disable such settings when he builds the releases.

JoshuaGreen commented:
I did a bit more testing on this, and I'm fairly confident that the differences I see can be ascribed to using somewhat larger memory allocations for each object.&nbsp; There are two factors at play here:
1. Extra memory is required to enforce alignment.
2. `BCMemValue` is one byte larger due to `BCMemValue::Leng` now being a `short`.

(1) can (now) be mitigated by manually setting the `MAX_ALIGNMENT` macro when compiling.&nbsp; While the `FXF` infrastructure should be completely generic, it's likely that on specific systems with our specific usage a smaller `MAX_ALIGNMENT` value than what's computed there will work.&nbsp; For example, on my Lubuntu virtual machine it seems that `MAX_ALIGNMENT` &ge;  8 is needed for well-definedness (according to [UBSan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html)) but even `MAX_ALIGNMENT == 1` doesn't seem to cause any obvious problems in practice.&nbsp; Experiments along these lines should be carefully tested before being used in specific releases.

Setting `MAX_ALIGNMENT` to 1 would still leave us with (2).&nbsp; Memory usage is still a bit higher, but obviously much closer to the original values, and the output I see aligns with every allocation being one-byte larger than before.

thomas-maeder commented:
This looks very good!
There are only 3 problems in all the *.inp files where we see a difference in the numbers produced if we set DOMEASURE:

```
hauntedchess.ref
323,326c323,326
<   add_to_move_generation_stack:   156050941
<                      play_move:   130481836
<  is_white_king_square_attacked:    34629453
<  is_black_king_square_attacked:   106030160
---
>   add_to_move_generation_stack:   167361760
>                      play_move:   134590349
>  is_white_king_square_attacked:    36197320
>  is_black_king_square_attacked:   110138673
nanna.ref
53,56c53,56
<   add_to_move_generation_stack:   699530479
<                      play_move:   703240780
<  is_white_king_square_attacked:    52542018
<  is_black_king_square_attacked:    26318548
---
>   add_to_move_generation_stack:   685036587
>                      play_move:   688631853
>  is_white_king_square_attacked:    55229450
>  is_black_king_square_attacked:    25636490
haan.out
301,304c301,304
<   add_to_move_generation_stack:   172423409
<                      play_move:   172423409
<  is_white_king_square_attacked:    43072625
<  is_black_king_square_attacked:   172169418
---
>   add_to_move_generation_stack:   183246656
>                      play_move:   183246656
>  is_white_king_square_attacked:    43113180
>  is_black_king_square_attacked:   182992665
```
It would be helpful to understand these differences before we merge the pull request.

thomas-maeder commented on DHT/dht.c:
```diff
@@ -877,21 +885,26 @@ void dhtRemoveElement(HashTable *ht, dhtConstValue key)
-  if (KeyV==0)
+  assert(key.value.object_pointer!=0); /* TODO: This assert assumes that object_pointer is the active member.
+                                          Is there a more generic test we could do?  Do we need one? */
+  if ((ht->procs.DupKeyValue)(key.value, &KeyV.value))
```
Hmm.
Before, we signaled "key duplication failed" if DupKey returned 0.
Now, we signal it if DupKeyValue doesn't.
Can this be correct?

JoshuaGreen replied:
It's admittedly confusing.&nbsp; Previously we `assert`ed that `key` is non-`NULL`, so now I `assert` that `key`'s stored object pointer is non-`NULL`.&nbsp; I've replaced the previous call to `DupKey` with a call to `DupKeyValue`, which is specified to return 0 on success and nonzero on failure (e.g., memory exhaustion), hence the nonzero return value corresponds to `DupKey` returning `NULL`.&nbsp; (`DupKeyValue` _is_ allowed [and supposed] to "duplicate" a `NULL` object pointer and return 0; this is needed elsewhere in this file.)

thomas-maeder commented on DHT/dht.c:
```diff
@@ -900,10 +913,9 @@ dhtElement *dhtEnterElement(HashTable *ht, dhtConstValue key, dhtConstValue data
   {
-    (ht->procs.FreeKey)((dhtValue)KeyV);
+    (ht->procs.FreeKeyValue)(KeyV.value);
     TraceText("data duplication failed\n");
```
Hmm.
Before, we signaled "data duplication failed" if DupData returned 0.
Now, we signal it if it doesn't.
Can this be correct?

JoshuaGreen replied:
Analogously to above, `DupData` is specified to return 0 on success and nonzero on failure, so we're still checking the analogous condition.&nbsp; What's special here was that previously we allowed for the original `data` to be `NULL` without signaling an error, hence I had to allow the same here.&nbsp; To make that work I had to allow `DupData` to simply copy a `NULL` pointer and return 0.

thomas-maeder commented on DHT/dhtmem.c:
```diff
@@ -16,66 +16,120 @@
     return 1;
+  while (length > ((size_t)-1))
+  {
+    if (memcmp(data1, data2, ((size_t)-1)))
```
The 3rd argument looks very strange.
Assuming that size_t is an unsigned type, this argument is very big.
Shouldn't we use the (equal) Length value of the two objects?

JoshuaGreen commented:
@thomas-maeder:

> It would be helpful to understand these differences before we merge the pull request.

Agreed, and this is what I'm currently investigating.&nbsp; This is a big update, and we shouldn't push it out until all concerns are resolved.&nbsp; Thanks for taking a first pass.

thomas-maeder commented:
I have reduced the haan chess example that produces a difference and ended up with this:

anfang
forderung       ser-#4
bedingung       HaanerSchach loch b1
option          Zugnummer
steine  weiss   kb3 16g6
        schwarz ka1
endeproblem

If I test this with -maxmem 15k, with DOMEASURE activated, I notice that the executable from this pull request prints higher numbers than the executable from branch develop.

The reason is that compresshash() is invoked after move
 #397 1939 16a5-g4 CURRMOVE_OF_PLY(nbply):13 nr_ghosts:0 current_ply:9
 . The executable from branch develop doesn't invoke compresshash().

compresshash() removes some elements from the hash table, which has the effect of the moves played to differ starting at move 2039:
< 2039 Kc4-c3 CURRMOVE_OF_PLY(nbply):13 nr_ghosts:0 current_ply:8
---
> 2039 Kb5-a4 CURRMOVE_OF_PLY(nbply):22 nr_ghosts:0 current_ply:9

I have not yet understood why one executable has to compress the hash table and the other doesn't. FWIW, the values of size computed in DupBCMemValue() seem to be equal.

NB: you changed the type of Leng in DHT/dhtbcmem.h to unsigned short (from unsigned char). Do we need Leng to be longer? Could its type be uint8_t?

JoshuaGreen commented:
There are a few things in this pull request that increase the sizes of allocated blocks.
1. Forcing alignment.
2. Bumping `BCMemValue::Leng` to an `unsigned short`.
3. (Using a `union` for `dhtValue` could as well, though that's less likely.)

At the time I believed (1) was necessary for correctness, and implementing it did eliminate some crashes that were otherwise unexplainable.&nbsp; Also, it's clear that we need to do this if we're going to pretend that `FXF` is a perfectly generic infrastructure.&nbsp; That written, I now suspect that the crashes were at least partially due to `dhtValue` originally including a `long double` member, as `long double`s have some particularly stringent alignment requirements.&nbsp; Having now removed that, we can likely get by with less on many systems.&nbsp; This can be experimented with by `#defin`ing `MAX_ALIGNMENT` at compile-time to a small power of 2.&nbsp; (In my next push that will be renamed to `FXF_MAX_ALIGNMENT`.)&nbsp; On my system the computed alignment is 16, but even setting `MAX_ALIGNMENT` to 1 doesn't seem to cause problems (though anything < 8 makes [UBSan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) complain).

(2) was implemented to resolve Issue #423.&nbsp; As @MuTsunTsai noted in PR #360, we could sometimes require more than 255 bytes to encode a position.&nbsp; My updates to [`optimisations/hash.h`](https://github.com/thomas-maeder/popeye/blob/feature/tighten_hash_usage_2/optimisations/hash.h) are intended to determine the maximum we could possibly need so that the initial buffers can be large enough.&nbsp; For now, it seems that an `unsigned short` is sufficient to store the length of any position encoding we may need.

JoshuaGreen commented:
As a sanity check, I tried running the non-lengthy regression testing with the alignment set to 1 and `BCMemValue::Leng` reverted back to an `unsigned char`.&nbsp; This eliminated two of the discrepancies, but `nanna.inp` discrepancies remain:
```
nanna.ref
53,56c53,56
<   add_to_move_generation_stack:   699530479
<                      play_move:   703240780
<  is_white_king_square_attacked:    52542018
<  is_black_king_square_attacked:    26318548
---
>   add_to_move_generation_stack:   700602407
>                      play_move:   704322485
>  is_white_king_square_attacked:    52565047
>  is_black_king_square_attacked:    26347819
```
I'll continue digging here.

JoshuaGreen commented:
This is interesting.&nbsp; I
- set the maximum alignment to 1
- set `BCMemValue::Leng` to `unsigned char`
- enabled `TESTHASH`
- (and eliminated the output between lines 981 and 990 of [`optimisations/hash.c`](https://github.com/thomas-maeder/popeye/blob/develop/optimisations/hash.c), and similarly in mine; do we _really_ need to keep this output around?)

I then tested each version on [`EXAMPLES/nanna.inp`](https://github.com/thomas-maeder/popeye/blob/develop/EXAMPLES/nanna.inp).&nbsp; The first difference appears after a compression when my version has two extra free blocks of size 27; all other free counts are the same, as are all malloced counts.&nbsp; The difference (though not the precise values) is the same after a call to `allocDHTelement`, but things look completely different before the next compression.&nbsp; This will take some digging.

JoshuaGreen commented:
For those who wish to help figure this out, I'd like to lay out my thoughts.&nbsp; I believe that many of the observed differences can be explained by my version simply allocating larger sizes than the current version.&nbsp; We can discuss the benefits of doing so, but I consider that secondary (until/unless we find a bug in that particular logic, of course).&nbsp; For now, the most important thing is to figure out why we see differences even when the allocations are the same size.&nbsp; To simulate this in my (current) code:
- Add `-DFXF_MAX_ALIGNMENT=1` to the compile line.
- Change `BCMemValue::Leng` (back) into an `unsigned char` (in [`DHT/dhtbcmem.h`](https://github.com/thomas-maeder/popeye/blob/feature/tighten_hash_usage_2/DHT/dhtbcmem.h)).

With these changes my `FXF` allocations should be identical (modulo which end of the overall allocation the blocks are taken from), and output should be identical.&nbsp; When _I_ do this I do see identical behavior except for [`EXAMPLES/nanna.inp`](https://github.com/thomas-maeder/popeye/blob/develop/EXAMPLES/nanna.inp), so understanding _those_ differences should be the first order of business (and is what I'm working on).

JoshuaGreen commented:
Progress.&nbsp; When in addition to the above I enforce `sizeof(void *)` alignment for any `FXF` allocations with sizes divisible by `sizeof(void *)` (like the current code does) then all the differences go away.&nbsp; This suggests that even x86 wants pointers aligned, and we've been lucky that, so far, none of our allocations that _aren't_ divisible by `sizeof(void *)` had any alignment requirements.&nbsp; That seems like an odd precondition for `FXF`, and we'll technically break it if we increase `BCMemValue::Leng` to `unsigned short`, though it's possible that that we will still be OK.&nbsp; That'll be my next test.

JoshuaGreen commented:
With the latest update, the first batch of differences have been resolved.&nbsp; That is, when I
- `$(DEFINEMACRO)FXF_MAX_ALIGNMENT=1` and
- revert `BCMemValue::Leng` (back) into an `unsigned char` (in [`DHT/dhtbcmem.h`](https://github.com/thomas-maeder/popeye/blob/feature/tighten_hash_usage_2/DHT/dhtbcmem.h))

then I no longer see _any_ differences in the non-lengthy test problems.&nbsp; I can now turn my attention to what happens when I keep `BCMemValue::Leng` as an `unsigned short` and/or resume enforcing alignment requirements.

Interestingly, though violating alignment requirements could _in principle_ cause anything to happen (though maybe not on x86, et al.), that wasn't what was causing the discrepancies.&nbsp; Instead, we were seeing effects of [_memory fragmentation_](https://en.wikipedia.org/wiki/Fragmentation_(computing)).&nbsp; As `FXF` blocks are allocated and freed, they break up the large allocated space into bunches of smaller portions.&nbsp; Larger blocks cannot be trivially constructed from these, so such freeing may not enable additional allocations to succeed.&nbsp; To minimize this, it helps to free objects in the opposite order that they were allocated in.&nbsp; [`DHT/dht.c`](https://github.com/thomas-maeder/popeye/pull/493/commits/bd7d0b5a72af2181413c7ae2a03100ca6b413b0a#diff-0dc003225deaa56971eb5e72b4e1bf1203eeb4c96d521828d0822b97f70cf128) mostly did this, but there were a couple of places where `InternHsElement`s, `dhtKey`s, and `dhtValue`s were allocated in one order but freed in another.&nbsp; I brought these cases into alignment; we now always allocate the `InternHsElement` first, followed by the `dhtKey` (if necessary), and then the `dhtValue`, and we free them in the reverse order.&nbsp; This minimizes memory fragmentation, and it was sufficient to eliminate the differences with the above "settings."

One may legitimately wonder why the above was only a problem for this code, why it apparently didn't show up in the current `develop` code or when I modified the code here to provide similar alignment guarantees (on top of `FXF_MAX_ALIGNMENT=1`).&nbsp; The reason is that the current `develop` code places allocations with sizes divisible by `sizeof(void *)` at the top of the (available) memory and other allocations at the bottom.&nbsp; This ensures alignment for the former, _and it also separates these different sets of allocations in memory_.&nbsp; The even-multiple allocations tend to be `InternHsElement`s while the latter are (necessarily) `dhtKey`s and/or `dhtValue`s, so it doesn't matter whether we free the `InternHsElement` or the `dhtKey`/`dhtValue` first; either way the given "portion" of memory (top or bottom) will see its most recent allocation freed before earlier allocations, preventing the fragmentation.&nbsp; Meanwhile, my simple change of `$(DEFINEMACRO)FXF_MAX_ALIGNMENT=1` in my code sent all allocations to the top, so they got in each other's way when freed.&nbsp; When I added alignment enforcement on top of `$(DEFINEMACRO)FXF_MAX_ALIGNMENT=1` the uneven sizes were forced to the bottom and everything worked again.

JoshuaGreen commented:
@thomas-maeder, could you please clarify the meanings of the
- add_to_move_generation_stack
- play_move
- is_white_king_square_attacked
- is_black_king_square_attacked

numbers?&nbsp; Specifically, if code changes could affect them, would we prefer to see them become lower, or higher?

thomas-maeder commented:
I have tried the current state of affairs with the following problem:

```
anfang
forderung       ser-#4
bedingung       HaanerSchach loch b1
option          Zugnummer
steine  weiss   kb3 16g6
        schwarz ka1
end
```

and command line py -maxmem 9k.

py generated from the sources of the development branch does the first compresshash() a bit later than py generated from this branch.

If I enable trace output in fxfAlloc(), I see the following:

branch develop:
```
fxfAlloc - size:288 sizeCurrentSeg:9216 current seg ptr-Arena:8928 MallocCount:1
fxfAlloc - size:2048 sizeCurrentSeg:8928 current seg ptr-Arena:6880 MallocCount:1
fxfAlloc - size:11 sizeCurrentSeg:6880 current seg ptr-Arena:0 MallocCount:1
fxfAlloc - size:24 sizeCurrentSeg:6865 current seg ptr-Arena:6856 MallocCount:1
fxfAlloc - size:13 sizeCurrentSeg:6841 current seg ptr-Arena:15 MallocCount:2
fxfAlloc - size:24 sizeCurrentSeg:6826 current seg ptr-Arena:6832 MallocCount:2
...

this branch:
```
fxfAlloc - size:288 sizeCurrentSeg:9216 current seg ptr-Arena:8928 MallocCount:1
fxfAlloc - size:2048 sizeCurrentSeg:8928 current seg ptr-Arena:6880 MallocCount:1
fxfAlloc - size:24 sizeCurrentSeg:6880 current seg ptr-Arena:0 MallocCount:1
fxfAlloc - size:12 sizeCurrentSeg:6856 current seg ptr-Arena:6864 MallocCount:1
fxfAlloc - size:24 sizeCurrentSeg:6840 FreeCount:1 current seg ptr-Arena:32 MallocCount:2
fxfAlloc - size:14 sizeCurrentSeg:6808 current seg ptr-Arena:6848 MallocCount:2
fxfAlloc - size:24 sizeCurrentSeg:6792 FreeCount:2 current seg ptr-Arena:64 MallocCount:3
```
Branch develop allocates 11 bytes for keys and 24 bytes for values.
This branch allocates 24 bytes for values and 12 bytes for keys.

In this branch, the allocations of key and value are in reverse order because you reordered the code, which I consider an improvement.

On the other hand, the added byte for the keys needs more analysis.

In branch develop, DupBCMemValue() computes size as sizeof *original  -  sizeof original->Data  +  original->Leng.
This branch computes size as offsetof(BCMemValue,Data)  +  original->Leng.

If I instrument function DupBCMemValue(), I get for the allocation of 11 bytes in branch develop:
DupBCMemValue: sizeof *original:2 sizeof original->Data:1 offsetof(BCMemValue,Data):1 original->Leng:10
In this branch, I get for the corresponding allocation of 12 bytes:
DupBCMemValue: sizeof *original:4 sizeof original->Data:1 offsetof(BCMemValue,Data):2 original->Leng:10

Once I change BCMemValue back to unsigned char in this branch, I get the same figures as for branch develop.

Branch develop still allocates slightly less though:
```
fxfAlloc - size:288 sizeCurrentSeg:9216 current seg ptr-Arena:8928 MallocCount:1
fxfAlloc - size:2048 sizeCurrentSeg:8928 current seg ptr-Arena:6880 MallocCount:1
fxfAlloc - size:11 sizeCurrentSeg:6880 current seg ptr-Arena:0 MallocCount:1
fxfAlloc - size:24 sizeCurrentSeg:6865 current seg ptr-Arena:6856 MallocCount:1
fxfAlloc - size:13 sizeCurrentSeg:6841 current seg ptr-Arena:15 MallocCount:2
fxfAlloc - size:24 sizeCurrentSeg:6826 current seg ptr-Arena:6832 MallocCount:2
fxfAlloc - size:13 sizeCurrentSeg:6802 current seg ptr-Arena:30 MallocCount:3
fxfAlloc - size:24 sizeCurrentSeg:6787 current seg ptr-Arena:6808 MallocCount:3
```
```
fxfAlloc - size:288 sizeCurrentSeg:9216 current seg ptr-Arena:8928 MallocCount:1
fxfAlloc - size:2048 sizeCurrentSeg:8928 current seg ptr-Arena:6880 MallocCount:1
fxfAlloc - size:24 sizeCurrentSeg:6880 current seg ptr-Arena:0 MallocCount:1
fxfAlloc - size:11 sizeCurrentSeg:6856 current seg ptr-Arena:6864 MallocCount:1
fxfAlloc - size:24 sizeCurrentSeg:6840 FreeCount:1 current seg ptr-Arena:32 MallocCount:2
fxfAlloc - size:13 sizeCurrentSeg:6808 current seg ptr-Arena:6848 MallocCount:2
fxfAlloc - size:24 sizeCurrentSeg:6792 FreeCount:2 current seg ptr-Arena:64 MallocCount:3
fxfAlloc - size:13 sizeCurrentSeg:6760 current seg ptr-Arena:6832 MallocCount:3
```
The minimum allocation size looks like 15 in branch develop, but 16 in this branch.
  fxfMINSIZE = sizeof(size_t)
vs.
  fxfMINSIZE = sizeof(void *)

Changing fxfMINSIZE to sizeof(size_t) (or even hardcoding it to 15) in this branch doesn't cause a difference.

The reason seems to be the new macro ALIGN_TO_MINIMUM.


I tend to think that we should leave the analysis at this. The impact of the more severe alignment is minimal, and aligning allocations to even boundaries seems to make sense.

JoshuaGreen commented:
@thomas-maeder, I'm inclined to agree that we now understand what's going on here.&nbsp; However, I'm not quite done.&nbsp; This code attempts to enforce possible alignment requirements, and that does come at some cost.&nbsp; I'm willing to trust you on what costs are acceptable, but it bothers me that the current `develop` code "incidentally" provides correct alignment for the types we're using while this code does not without accepting those costs.&nbsp; Therefore, I'd like to split the `FXF_MAX_ALIGNMENT` macro into (at least) two separate macros that we can set when building to ensure that
- allocations with sizes divisible by `void *` are aligned to `void *`'s alignment, and
- other allocations are aligned to `unsigned short`'s alignment, what we'll now need for `BCMemValue`.

Once I've set this up, we should be able to ensure the alignment we need (for well-defined behavior) while keeping the infrastructure fully general and minimizing the costs.

JoshuaGreen commented:
I'm ready to open this branch up to independent testing and, perhaps, eventual merging.

I restored the previous behavior of putting even allocations at the top of the range and uneven ones at the bottom.&nbsp; Though the `FXF` code will _by default_ ensure sufficient alignment for any use, I now allow a bit of customization that, admittedly, is _very specific_ to Popeye's precise usage (and which the old version implicitly relied on anyway).&nbsp; Specifically, I allow compile-time macros to set the alignment(s) needed by specifying the relevant types.&nbsp; In the build system I've set those to types that I think are appropriate, and the remaining differences in testing are minimal.

There is the downside that it's impossible to verify that these settings are correct across all systems or that they will remain correct as the code continues to evolve.&nbsp; As a mitigation, I modified the `fxfAlloc` interface to take the _type_ as an argument, and in debug mode that type's alignment will be compared to the alignment `FXF` intends to provide.&nbsp; If the needed alignment is larger than the provided alignment then a warning will be emitted.&nbsp; This warning is non-fatal, but it should hopefully allow us to detect and eliminate any misalignment issues before or when they cause problems on some system.

thomas-maeder commented:
Hi Joshua

After merging branch develop into feature/tighten_hash_usage_2, I have thrown the result at all the .inp files;  no significant difficulties have shown up.
I have therefore merged this pull request and deleted branch feature/tighten_hash_usage_2.

Can you verify if the open issues you hoped would be resolved by this have indeed been resolved?
